### ISSUE TITLE
Convert PullRequest builds to be "pipeline of jobs"

### ISSUE THREAD

**Issue by AdamBrousseau** [eclipse/openj9#2728]
Related #1450 

Currently our PR builds execute in a single job. We use our OpenJ9 build AND test pipeline scripts in order to complete the build. This is inefficient because it is a different process than how all our other builds are now working. In order to complete #1450 and remove our test pipeline code, we need to convert the PR builds to use the [Pipeline-Build-Test-Any-Platform](https://github.com/eclipse/openj9/blob/master/buildenv/jenkins/jobs/pipelines/Pipeline-Build-Test-Any-Platform). This will also allow for other PR work to proceed. There are two parts in order to get this working
1. Reconfig the PR jobs to use the above file instead of the matching PR files
1a. Add the PLATFORM, SDK_VERSION and TESTS_TARGETS (not TEST_TARGET or TEST_TARGETS) parameters to each PR job
1. Modify the pipeline files ([1](https://github.com/eclipse/openj9/blob/master/buildenv/jenkins/common/pipeline-functions#L218)) to pass the PR related parameters we need in the build.
   - ghprbPullId
   - ghprbGhRepository
   - ghprbCommentBody
   - ghprbTargetBranch

/cc @llxia @jdekonin @vsebe

**Comment by AdamBrousseau** [comment]
Summarizing here since the PR is pretty messy at this point. I have booked a meeting next week to discuss this. Hopefully we can get this resolved soon.

#### Advantages
- Use Adopt Test Pipelines
   - No longer maintain our own test pipeline scripts
- Align with the same build & test story as our nightly, omr and personal builds.
- Allows for SVT and other testing in PRs (JitaaS etc)
- Reduced master disk usage. See **Concerns**.
- Reduced releng maintenance
   - No complicated regexs
   - Less builds to duplicate each new java release
   - No dup test pipelines to maintain/debug
- Unblocks future work
   - Stale pipeline code cleanup #2753
   - Grinder-style PR builds (no compiles) #2859
   - Move Jenkins params to pipeline code for PRs #2744
- Reduced overall machine time
   - 1 Compile per build rather than per test

#### Disadvantages
- Overall build time will likely increase in most scenarios.
   - Machines will be less busy, so when the farm is loaded builds may get through faster. But when the farm is idle the build time will be increased by the amount of time to transfer the SDK(s) from the compile machine to the Jenkins master and then back down to the test machine.
   - See numbers [here](https://github.com/eclipse/openj9/pull/2836#issuecomment-439212819)
- Cannot test pipeline code changes in a PR build. Currently the builds check out the repo at the merge commit of the PR which means the pipeline code you get/run has your changes. When we change over, you'll only get master pipeline code. This includes the defaults.yml variable file. I believe there may be a way to pass the scm details along but I'll leave that for follow up issue.

#### Follow-up improvements
- How to test pipeline changes in PR builds #4443
- Local SDK Caches
- Build logs #3573 
- Artifact storage #2637

#### Concerns
 - Disk usage
   - I have done some inventory and calculations based on our current usage. We should reduce usage with this change. Numbers [here](https://github.com/eclipse/openj9/pull/2836#issuecomment-424945878) and [here](https://github.com/eclipse/openj9/pull/2836#issuecomment-447020119)

- Auto job cancellation
   - Currently we have a very granular level of "auto-cancel" on the PR builds. Per spec, per version, per test target. I have mimicked this behaviour in the new world. When the orchestration job (Pipeline-Build-test-VERSION-SPEC) attempts to launch a test build, it first cancels any of the same test job (per spec, per version, per target), then launches the new job. What it will not auto-cancel is compile jobs. This is because it would be much more tedious to determine "what is coming after the compile" to know whether or not it should be cancelled. This will allow for some possible machine wasted usage. Jobs can also be cancelled manually by any committer, although it is not expected this will happen.

- Build Times
   - See Disadvantages

- UNB Network
   - Based on my [testing](https://github.com/eclipse/openj9/pull/2836#issuecomment-457335472) I do not see any network perf implications of running 12 jobs in parallel at UNB.

**Comment by jdekonin** [comment]
Additional follow up improvements
- I discussed with @charliegracie a request for increasing the throughput at UNB.  its believed we have a 100MB line
- log improvements within the test pipelines; maybe not keep any logs from a passing run with an option to control

**Comment by DanHeidinga** [comment]
**Rough notes from the meeting:**

Questions:
* History:
  * Can we use different jobs to launch the same pipeline so that we maintain ~5 nights worth of nightly build history?
  * Rationale: nightly build issues may not be noticed as quickly as PR failures.
  * Provides separate history for nightly, omr, and prs.
  * Makes it easier to find the nightlies vs other build types

* Disk usage
  * Can modify the "discard builds" settings to help alleviate the disk usage increase concerns

* Typical use case:
  * multiple comments to launch builds: sanity on one platform, extended on another, compile on a third, etc
     * will add new compile / test children to the same top-level comment - will have access to all the launched jobs
  * repeating a command used to cancel the previous build
     * Now, test job will cancel the previous test jobs, but not the compiles
     * Compiles are not canceled.  Previous compile may (in worst case) finish out of order.

* PR build times will longer and typically slower as they will require two machines and upload / download for the SDK.
  * Save compiles so potentially reduce the overall farm load
  * Adopt builds are looking at enabling parallel testing

* Upload / download times 
  * upload times are between 1-2 minutes.  zLinux is ~10 min
  * download times range from 1 to 40 minute.  The 40 min downloads happen across platforms
     * distribution on the times?  **Adam to get a data**
     * Are some platforms worse than others?

* Any files in the `buildenv/jenkins`can't be tested in a PR build.  Will need email to let committers know we can't test these files in a PR build.  Use sandbox side jobs?

* UNB need to determine if we share a pipe with the UNB CS servers and how to measure whether we're over saturating the link

**Comment by pshipton** [comment]
Stated early, but also note that PR build times can be faster when only targeted testing is required. i.e. in the future a PR build could run only the CryptoTest instead of all sanity testing, for a change that only affects this test

**Comment by harryyu1994** [comment]
Hi, the JITaaS team is wondering when this work will be finished. The JITaaS PR testing feature is blocked by this. We are in no rush to have this feature now and just asking about the timeline. Thanks!

**Comment by AdamBrousseau** [comment]
Artifactory work https://github.com/eclipse/openj9/issues/2637#issuecomment-463713111

**Comment by AdamBrousseau** [comment]
Remaining concerns

>History -  Can we use different jobs to launch the same pipeline so that we maintain ~5 nights worth of nightly build history?

- [ ] Separate builds by build identifier. Can be done now (ahead of time). Will allow for granular tuning of amount of history for each job.
   - [ ] Autogen jobs #4782 will alleviate concerns over job creation
   - [ ] Update variable file build_discarder with values for each build
   - [ ] Email dev mailing list

> Disk usage

> PR build times will longer and typically slower as they will require two machines and upload / download for the SDK.

- [ ] Enable Artifactory #2637. Moves all the artifacts off of master and only leaves logs to consume disk space. This also improves transfer times for artifacts in builds.
   - [x] Determine access https://github.com/eclipse/openj9/issues/2637#issuecomment-465280056
   - [ ] HTTPS Certificate [Bug 541852](https://bugs.eclipse.org/bugs/show_bug.cgi?id=541852#c7)
   - [ ] Cleanup script/job (non-blocking)
   - [ ] PR for Openj9 #4705
   - [ ] PR for Test AdoptOpenJDK/openjdk-tests#898

> * Any files in the `buildenv/jenkins`can't be tested in a PR build.  Will need email to let committers know we can't test these files in a PR build.  Use sandbox side jobs?

- [ ] Email dev mailing list re testing Pipeline changes in new world. Can be part of "switch over" email
- [ ] Determine how to test pipe changes in PRs (non-blocking #4443)


#### Turning on
Once Artifactory is enabled and jobs are separated by build identifier,
- Email dev list
- Merge #2836 
- Disable all PullRequest-[Compile|Sanity|Extended]* jobs
- Enable PullRequest-[OpenJ9|OpenJDK8|OpenJDK11|OpenJDK12|OpenJDK]

**Comment by DanHeidinga** [comment]
@AdamBrousseau If a problem is discovered after making the switch, is it easy to backout of the change?  What would be involved?

**Comment by AdamBrousseau** [comment]
Yes, it would simply be the reverse of that list.

**Comment by DanHeidinga** [comment]
Is it essential to revert #2836 or is toggling the enable / disable of the jobs sufficient?

**Comment by AdamBrousseau** [comment]
Last time I tested it, the changes in 2836 did not work for both worlds. I didn't determine why at the time. I can re-look at this and retest it.

**Comment by AdamBrousseau** [comment]
>Last time I tested it, the changes in 2836 did not work for both worlds. I didn't determine why at the time. I can re-look at this and retest it.

Seems to work now with my changes for the old and the new world. So we could merge this ahead of time and use the builds to flip back/forth as needed.

**Comment by AdamBrousseau** [comment]
This is delivered and setup as default. The remaining task is #4443.
